### ISSUE TITLE
gringo fails to build in master

### ISSUE THREAD

**Issue by Zimmi48** [NixOS/nixpkgs#19950]
gringo-4.5-4 fails to build in the master branch

@cleverca22 helped me find out the commit creating the problem. It is f42a7068.
gringo depends on scons and for some reason the update of scons broke gringo.
Tagging @FRidh as well since he is the author of the commit.

If I understood the trace correctly, the error reported is `call of overloaded 'to_string(unsigned int)' is ambiguous`.
It happens in a function defined as such:

```
FWString AuxGen::uniqueName(char const *prefix) {
    return FWString(prefix + std::to_string((*auxNum)++));
}
```

and @cleverca22 suggests to patch this definition in the following way

```
FWString AuxGen::uniqueName(char const *prefix) {
    unsigned long x = (*auxNum)++;
    return FWString(prefix + std::to_string(x));
}
```

It still remains to understand how this commit could have created this problem...

**Comment by FRidh** [comment]
Building against that specific commit is indeed unsuccessful:
```
nix-shell -p gringo -I nixpkgs=https://github.com/NixOS/nixpkgs/archive/f42a7068da8956a86a4e71018022222d63cf2b7a.tar.gz
```
however, building against current `nixpkgs-unstable` (de3860229ae1dee77a2e04015e590801befafa25) , which is more recent, works
```
nix-shell -p gringo -I nixpkgs=https://github.com/NixOS/nixpkgs-channels/archive/nixpkgs-unstable.tar.gz
```
Yet building against master fails again.

**Comment by FRidh** [comment]
After undoing the commit the build of `gringo` is successful again, so it definitely seems `scons` related. None of the other packages that use `scons` fails building though. 

The documented method to install was how it was before, although nowadays all Python packages are build as a wheel and then installed. Somehow something is different there. I did remove two options in that commit, one which was a compiler optimization. When I do pass those, the error still occurs. 

If the patch makes it build again, I suggest using it and reporting it upstream.

**Comment by Zimmi48** [comment]
I have not tested the patch.

Le Ven 28 Oct 2016 19:38, Frederik Rietdijk <notifications@github.com> a
écrit :

**Comment by sternenseemann** [comment]
I have tested the patch just now and it doesn't fix the problem.

**Comment by FRidh** [comment]
`gringo` seems to be discontinued so there doesn't seem to be a place to ask if they've encountered the same issue. Note that `gringo` has been superseded by [clingo](https://github.com/potassco/clingo).

Since all other packages depending on `scons` build fine, I would prefer not to undo the commit. Therefore, could you copy the old expression of `scons` into your expression of gringo, and use that version of `scons` to build `gringo`?

**Comment by sternenseemann** [comment]
Can we replace `gringo` with `clingo` trivially?

**Comment by sternenseemann** [comment]
This derivation, as suggested by @FRidh, works, but I can't work out why this is any different from the usual scons which really bothers me. It doesn't seem like a good fix for the problem.

```nix
{ stdenv, fetchurl,
  bison, re2c,
  python, makeWrapper
}:

let
  scons = stdenv.mkDerivation rec {
    version = "2.5.0";
    name = "scons-${version}";

    src = fetchurl {
      url = "mirror://sourceforge/scons/${name}.tar.gz";
      sha256 = "1sdcj8iapfzmlq7098yas40qwl6khsiwydbxv33sw81wy93nnagb";
    };

    buildInputs = [python makeWrapper];

    preConfigure = ''
    for i in "script/"*; do
    substituteInPlace $i --replace "/usr/bin/env python" "${python}/bin/python"
    done
    '';
    buildPhase = "python setup.py install --prefix=$out --install-data=$out/share --install-lib=$(toPythonPath $out) --symlink-scons -O1";
    installPhase = "for n in $out/bin/*-${version}; do wrapProgram $n --suffix PYTHONPATH ':' \"$(toPythonPath $out)\"; done";

    pythonPath = [];
  };

in

stdenv.mkDerivation rec {
  name = "gringo-${version}";
  version = "4.5.4";

  src = fetchurl {
    url = "mirror://sourceforge/project/potassco/gringo/${version}/gringo-${version}-source.tar.gz";
    sha256 = "16k4pkwyr2mh5w8j91vhxh9aff7f4y31npwf09w6f8q63fxvpy41";
  };

  buildInputs = [ bison re2c scons ];

  patches = [
    ./gringo-4.5.4-cmath.patch
  ];

  buildPhase = ''
    scons --build-dir=release
  '';

  installPhase = ''
    mkdir -p $out/bin
    cp build/release/gringo $out/bin/gringo
  '';

  meta = with stdenv.lib; {
    description = "Converts input programs with first-order variables to equivalent ground programs";
    homepage = http://potassco.sourceforge.net/;
    platforms = platforms.linux;
    maintainers = [ maintainers.hakuch ];
    license = licenses.gpl3Plus;
  };
}
```

**Comment by Zimmi48** [comment]
Personally, I'm not specifically interested in `gringo`. I would like to build OCaml packages which depend on `opam`. `opam`'s derivation depends on `aspcud` which in turns depends on `gringo`.
Here is an explanation from the OPAM team of the dependency on `aspcud`: http://opam.ocaml.org/doc/Install.html#ExternalSolvers. It could apparently be dropped completely or replaced with another solver such as `packup`.

`aspcud` is from the same team as `gringo` and `clingo`. From this page http://www.cs.uni-potsdam.de/wv/aspcud/ it seems that `aspcud` depends on `gringo` but `gringo` itself is now part of `clingo`, as shown on https://github.com/potassco/clingo:

> Clingo combines both gringo and clasp into a monolithic system.

Additionally, it looks like `aspcud` is the only package depending on `gringo` in nixpkgs: https://github.com/NixOS/nixpkgs/search?q=gringo
and `opam` is the only package depending on `aspcud`: https://github.com/NixOS/nixpkgs/search?q=aspcud

To me any solution would be fine (replacing `gringo` with `clingo` or replacing `aspcud` with something else or merely dropping this dependency). Maybe the OCaml team of nixpkgs should be consulted on this matter and `topic: ocaml` be added to this issue.

**Comment by sternenseemann** [comment]
Same applies for me, I ran into it as well, when trying to package `ptime`.

**Comment by FRidh** [comment]
I leave finding a solution up to you, but, if you need a quick fix, then just open a PR as I proposed and that would be fine.

**Comment by Zimmi48** [comment]
It looks easy to replace `gringo` by `clingo` so I'm going to try that and if it works, I'll open a PR.

**Comment by Zimmi48** [comment]
Closed by #20086
